File: scripts/test-create-template.js
#!/usr/bin/env node
/**
 * Test template creation by calling dashboard functions directly.
 * No pexpect, no TUI — tests the actual code paths.
 */
const fs = require('fs');
const path = require('path');

const TEMPLATES_DIR = path.join(__dirname, '..', 'templates');
const tplName = 'test-character-' + Date.now();
const tplDir = path.join(TEMPLATES_DIR, tplName);

let passed = 0, failed = 0;
function check(label, ok, detail) {
  if (ok) { console.log(`  ✅ ${label}`); passed++; }
  else { console.log(`  ❌ ${label}${detail ? ': ' + detail : ''}`); failed++; }
}

async function main() {
  console.log('\n╔══════════════════════════════════════════════════╗');
  console.log('║  Template Creation Unit Test                      ║');
  console.log('╚══════════════════════════════════════════════════╝\n');

  // ── Test 1: Template name normalization ──
  console.log('Test 1: Name normalization\n');
  check('"shreck character" → "shreck-character"', 'shreck character'.toLowerCase().replace(/\s+/g, '-').replace(/[^a-z0-9-]/g, '') === 'shreck-character');
  check('"My Agent!!" → "my-agent"', 'My Agent!!'.toLowerCase().replace(/\s+/g, '-').replace(/[^a-z0-9-]/g, '') === 'my-agent');
  check('"code review" → "code-review"', 'code review'.toLowerCase().replace(/\s+/g, '-').replace(/[^a-z0-9-]/g, '') === 'code-review');
  check('"---test---" → "---test---"', '---test---'.toLowerCase().replace(/\s+/g, '-').replace(/[^a-z0-9-]/g, '') === '---test---');
  check('empty → rejected', ''.toLowerCase().replace(/\s+/g, '-').replace(/[^a-z0-9-]/g, '') === '');

  // ── Test 2: Build config.json structure ──
  console.log('\nTest 2: Config structure\n');

  const config = {
    template: tplName,
    profile: {
      name: 'Test Character',
      type: 'autonomous',
      description: 'A test character for template validation',
      network: {
        capabilities: ['storytelling', 'roleplay'],
        protocols: ['MCP', 'REST'],
      },
      profile: {
        category: 'entertainment-gaming',
        tags: ['RPG', 'character', 'test'],
      },
      models: ['moonshotai/kimi-k2.5'],
      markup: 15,
      session: { duration: 3600, tokenLimit: 100000, messageLimit: 50 },
      workspace: { enabled: false },
    },
    service: {
      name: 'Test Character Chat',
      description: 'Chat with a test character',
      price: '0.005',
      currency: 'VRSCTEST',
      category: 'entertainment-gaming',
      turnaround: 'instant',
      paymentTerms: 'prepay',
      sovguard: true,
    },
  };

  check('Config has template name', config.template === tplName);
  check('Config has profile.name', config.profile.name === 'Test Character');
  check('Config has profile.type', ['autonomous', 'assisted', 'hybrid', 'tool'].includes(config.profile.type));
  check('Config has profile.description', config.profile.description.length > 10);
  check('Config has capabilities array', Array.isArray(config.profile.network.capabilities));
  check('Config has protocols array', Array.isArray(config.profile.network.protocols));
  check('Config has tags array', Array.isArray(config.profile.profile.tags));
  check('Config has category string', typeof config.profile.profile.category === 'string');
  check('Config has models array', Array.isArray(config.profile.models));
  check('Config has markup number', typeof config.profile.markup === 'number');
  check('Config has session params', config.profile.session.duration > 0);
  check('Config has workspace', typeof config.profile.workspace.enabled === 'boolean');
  check('Config has service.name', config.service.name.length > 0);
  check('Config has service.price', parseFloat(config.service.price) > 0);
  check('Config has service.currency', config.service.currency.length > 0);
  check('Config has service.sovguard', typeof config.service.sovguard === 'boolean');
  check('Config has service.paymentTerms', ['prepay', 'postpay', 'split'].includes(config.service.paymentTerms));

  // ── Test 3: Build SOUL.md ──
  console.log('\nTest 3: SOUL.md builder\n');

  const role = 'You are Shreck, the ogre who lives in a swamp';
  const personality = 'Grumpy but lovable, Scottish accent';
  const rules = 'Never break character, never say you are an AI';
  const style = 'Short grumpy sentences with ogre metaphors';
  const catchphrases = 'What are ye doin in me swamp, Ogres have layers';
  const extra = 'You secretly care about people';

  const lines = [role];
  if (personality) lines.push('', 'Personality: ' + personality);
  if (rules) lines.push('', 'Rules:', ...rules.split(',').map(r => '- ' + r.trim()));
  if (style) lines.push('', 'Style: ' + style);
  if (catchphrases) lines.push('', 'Key phrases:', ...catchphrases.split(',').map(p => '- "' + p.trim() + '"'));
  if (extra) lines.push('', extra);
  const soulPrompt = lines.join('\n');

  check('SOUL has role', soulPrompt.includes('Shreck'));
  check('SOUL has personality', soulPrompt.includes('Scottish'));
  check('SOUL has rules', soulPrompt.includes('Never break character'));
  check('SOUL has style', soulPrompt.includes('ogre metaphors'));
  check('SOUL has catchphrases', soulPrompt.includes('swamp'));
  check('SOUL has extra', soulPrompt.includes('secretly care'));
  check('SOUL multi-line', soulPrompt.split('\n').length > 5);

  // ── Test 4: Save template ──
  console.log('\nTest 4: Save template\n');

  fs.mkdirSync(tplDir, { recursive: true });
  fs.writeFileSync(path.join(tplDir, 'config.json'), JSON.stringify(config, null, 2));
  fs.writeFileSync(path.join(tplDir, 'SOUL.md'), soulPrompt);

  check('Template dir created', fs.existsSync(tplDir));
  check('config.json saved', fs.existsSync(path.join(tplDir, 'config.json')));
  check('SOUL.md saved', fs.existsSync(path.join(tplDir, 'SOUL.md')));

  // Verify roundtrip
  const loaded = JSON.parse(fs.readFileSync(path.join(tplDir, 'config.json'), 'utf8'));
  check('config.json roundtrip', loaded.template === tplName);
  check('config.json profile intact', loaded.profile.name === 'Test Character');
  check('config.json service intact', loaded.service.name === 'Test Character Chat');

  const loadedSoul = fs.readFileSync(path.join(tplDir, 'SOUL.md'), 'utf8');
  check('SOUL.md roundtrip', loadedSoul.includes('Shreck'));

  // ── Test 5: Template appears in list ──
  console.log('\nTest 5: Template discovery\n');

  const templates = fs.readdirSync(TEMPLATES_DIR).filter(d =>
    fs.existsSync(path.join(TEMPLATES_DIR, d, 'config.json'))
  );
  check('Template in directory listing', templates.includes(tplName));
  check('All templates valid', templates.every(t =>
    fs.existsSync(path.join(TEMPLATES_DIR, t, 'config.json'))
  ));

  // ── Test 6: Template works with `setup --template` ──
  console.log('\nTest 6: Template compat with setup command\n');

  // Check which fields setup actually reads from the template
  const setupFields = ['name', 'type', 'description'];
  for (const field of setupFields) {
    check(`profile.${field} present`, loaded.profile[field] != null);
  }
  check('profile.profile.category present', loaded.profile.profile?.category != null);
  check('profile.profile.tags present', Array.isArray(loaded.profile.profile?.tags));
  check('profile.network.protocols present', Array.isArray(loaded.profile.network?.protocols));
  check('service.name present', loaded.service?.name != null);
  check('service.price present', loaded.service?.price != null);

  // ── Cleanup ──
  fs.rmSync(tplDir, { recursive: true });
  check('Cleanup', !fs.existsSync(tplDir));

  // ── Summary ──
  console.log(`\n══════════════════════════════════════════════════`);
  console.log(`  Results: ${passed} passed, ${failed} failed`);
  console.log(`══════════════════════════════════════════════════\n`);

  process.exit(failed > 0 ? 1 : 0);
}

main();

File: scripts/test-create-template.py
#!/usr/bin/env python3
"""
Test the full Create Custom Template flow via the interactive dashboard.
Creates a "character-roleplay" template with all fields filled out.
"""
import pexpect
import sys
import time
import os

os.chdir(os.path.join(os.path.dirname(__file__), '..'))

TIMEOUT = 30
DOWN = '\x1b[B'
UP = '\x1b[A'
ENTER = '\r'
ESC = '\x1b'

passed = 0
failed = 0

def check(label, ok, detail=''):
    global passed, failed
    if ok:
        print(f'  ✅ {label}')
        passed += 1
    else:
        print(f'  ❌ {label}{": " + detail if detail else ""}')
        failed += 1

import re
ANSI_RE = re.compile(r'\x1b\[[0-9;]*[a-zA-Z]|\x1b\[\?[0-9]*[a-zA-Z]')

def wait_for(p, pattern, timeout=TIMEOUT):
    """Wait for pattern in output, ignoring ANSI escape codes"""
    end_time = time.time() + timeout
    while time.time() < end_time:
        try:
            p.expect('.+', timeout=1)
            # Strip ANSI from accumulated buffer
            clean = ANSI_RE.sub('', p.before + p.after if p.after else p.before)
            if re.search(pattern, clean):
                return True
        except pexpect.TIMEOUT:
            continue
        except:
            return False
    return False

def nav_down_enter(p, n):
    for _ in range(n):
        p.send(DOWN)
        time.sleep(0.08)
    time.sleep(0.15)
    p.send(ENTER)
    time.sleep(0.5)

def type_text(p, text):
    time.sleep(0.3)  # Wait for prompt to be ready before typing
    p.sendline(text)
    time.sleep(0.5)

def toggle_checkbox(p, positions):
    """For checkbox prompts — navigate to positions and toggle with space"""
    for pos in positions:
        for _ in range(pos):
            p.send(DOWN)
            time.sleep(0.08)
        p.send(' ')  # space to toggle
        time.sleep(0.1)
    p.send(ENTER)
    time.sleep(0.5)


print('\n╔══════════════════════════════════════════════════╗')
print('║  Create Custom Template — Full Flow Test          ║')
print('╚══════════════════════════════════════════════════╝\n')

# Clean up any previous test template
template_dir = 'templates/character-roleplay'
if os.path.exists(template_dir):
    import shutil
    shutil.rmtree(template_dir)
    print('  (cleaned up previous template)\n')

p = pexpect.spawn('node src/cli.js', encoding='utf-8', timeout=TIMEOUT, dimensions=(60, 120))
# p.logfile_read = sys.stdout  # uncomment for debug

# ── Step 1: Navigate to Add New Agent ──
print('Step 1: Navigate to Add New Agent\n')
check('Banner', wait_for(p, 'J41 Dispatcher', 15))
wait_for(p, 'What would you like to do', 10)
nav_down_enter(p, 1)  # [2] Add New Agent
check('Add Agent screen', wait_for(p, 'Add New Agent', 10))

# ── Step 2: Fill agent ID and name ──
print('\nStep 2: Agent ID and name\n')
check('Agent ID prompt', wait_for(p, 'Agent ID', 10))
type_text(p, 'agent-template-test')
check('Identity prompt', wait_for(p, 'Identity name|identity|name', 15))
type_text(p, 'templatetest1')
check('Template select', wait_for(p, 'Select template|template', 15))

# ── Step 3: Select "+ Create Custom Template" (last item after separator) ──
print('\nStep 3: Select Create Custom Template\n')
# Navigate to bottom — custom template is after the 3 existing templates + separator
# Navigate past existing templates to "+ Create Custom Template"
# 4 templates now (code-review, data-analyst, general-assistant, workspace-reviewer) + separator + custom
nav_down_enter(p, 5)
check('Custom template screen', wait_for(p, 'Create Custom Template|Template name', 10))

# ── Step 4: Template name ──
print('\nStep 4: Template name\n')
wait_for(p, 'Template name', 10)
type_text(p, 'character-roleplay')
time.sleep(1)  # Wait for normalization message + screen redraw
check('Name entered', wait_for(p, 'Display name|Profile|name', 15))

# ── Step 5: Agent Profile ──
print('\nStep 5: Agent Profile\n')

# Display name
wait_for(p, 'Display name', 10)
type_text(p, 'Shreck the Ogre')
check('Display name', True)

# Agent type — select autonomous (first item, default)
wait_for(p, 'Agent type', 10)
p.send(ENTER); time.sleep(0.5)
check('Agent type', True)

# Description
wait_for(p, 'Description', 10)
type_text(p, 'Visit Shreck in his swamp — character roleplay agent')
check('Description', True)

# Category — select from list
wait_for(p, 'Category', 10)
# Need to find "Entertainment & Gaming" — scroll down
for i in range(20):
    p.send(DOWN)
    time.sleep(0.08)
    # Check if we can see entertainment
# Just pick wherever we are for now — we'll verify the template file
p.send(ENTER); time.sleep(0.5)
check('Category selected', True)

# Tags
wait_for(p, 'Tags', 10)
type_text(p, 'RPG,roleplay,character,shreck')
check('Tags', True)

# Markup
wait_for(p, 'Markup', 10)
type_text(p, '15')
check('Markup', True)

# Models
wait_for(p, 'Models', 10)
type_text(p, 'moonshotai/kimi-k2.5')
check('Models', True)

# Protocols — checkbox, MCP and REST should be default selected
wait_for(p, 'Protocols', 10)
p.send(ENTER); time.sleep(0.5)  # Accept defaults
check('Protocols', True)

# Capabilities
wait_for(p, 'Capabilities', 10)
type_text(p, 'storytelling,roleplay,conversation')
check('Capabilities', True)

# Workspace — No
wait_for(p, 'workspace|Enable', 10)
# Navigate to "No" — for confirm prompts, type 'n'
type_text(p, 'n')
check('Workspace disabled', True)

# ── Step 6: Session Limits ──
print('\nStep 6: Session Limits\n')

wait_for(p, 'duration|Session', 10)
type_text(p, '3600')
check('Duration', True)

wait_for(p, 'Token limit', 10)
type_text(p, '100000')
check('Token limit', True)

wait_for(p, 'Message limit', 10)
type_text(p, '50')
check('Message limit', True)

# ── Step 7: Service ──
print('\nStep 7: Default Service\n')

wait_for(p, 'Service name', 10)
type_text(p, 'Shreck Swamp Visit')
check('Service name', True)

wait_for(p, 'Service description', 10)
type_text(p, 'Chat with Shreck in his swamp')
check('Service description', True)

wait_for(p, 'Price', 10)
type_text(p, '0.005')
check('Price', True)

wait_for(p, 'Currency', 10)
type_text(p, 'VRSCTEST')
check('Currency', True)

# Service category
wait_for(p, 'Category', 10)
for i in range(20):
    p.send(DOWN)
    time.sleep(0.08)
p.send(ENTER); time.sleep(0.5)
check('Service category', True)

wait_for(p, 'Turnaround', 10)
type_text(p, 'instant')
check('Turnaround', True)

# Payment terms — list
wait_for(p, 'Payment terms', 10)
p.send(ENTER); time.sleep(0.5)  # prepay (default)
check('Payment terms', True)

# SovGuard — yes
wait_for(p, 'SovGuard', 10)
p.send(ENTER); time.sleep(0.5)  # default yes
check('SovGuard', True)

# ── Step 8: SOUL.md ──
print('\nStep 8: SOUL.md\n')

wait_for(p, 'SOUL|personality|prompt', 10)
# Select "Write custom personality" (second option)
nav_down_enter(p, 1)
time.sleep(1)
check('Custom SOUL selected', wait_for(p, 'Who is this|agent|role', 10))

type_text(p, 'You are Shreck, the ogre who lives in a swamp')
check('Role', True)

wait_for(p, 'Personality', 10)
type_text(p, 'Grumpy but lovable, Scottish accent, loves his swamp')
check('Personality', True)

wait_for(p, 'Rules', 10)
type_text(p, 'Never break character, never say you are an AI, always mention the swamp')
check('Rules', True)

wait_for(p, 'Communication style', 10)
type_text(p, 'Short grumpy sentences with ogre metaphors about onions and layers')
check('Style', True)

wait_for(p, 'Key phrases', 10)
type_text(p, 'What are ye doin in me swamp, Ogres have layers, Better out than in')
check('Catchphrases', True)

wait_for(p, 'Anything else', 10)
type_text(p, 'You secretly care about people but hide it behind grumpiness')
check('Extra', True)

# Preview + confirm
wait_for(p, 'Use this personality', 10)
p.send(ENTER); time.sleep(0.5)  # Yes
check('SOUL confirmed', True)

# ── Step 9: Template saved ──
print('\nStep 9: Verify template saved\n')
time.sleep(2)

# Check the template was saved
if os.path.exists(template_dir):
    check('Template dir created', True)
    check('config.json exists', os.path.exists(os.path.join(template_dir, 'config.json')))
    check('SOUL.md exists', os.path.exists(os.path.join(template_dir, 'SOUL.md')))

    # Verify config.json content
    import json
    try:
        cfg = json.load(open(os.path.join(template_dir, 'config.json')))
        check('Template name', cfg.get('template') == 'character-roleplay')
        check('Profile name', cfg.get('profile', {}).get('name') == 'Shreck the Ogre')
        check('Profile type', cfg.get('profile', {}).get('type') == 'autonomous')
        check('Has service', cfg.get('service', {}).get('name') == 'Shreck Swamp Visit')
        check('Service price', cfg.get('service', {}).get('price') == '0.005')
        check('SovGuard enabled', cfg.get('service', {}).get('sovguard') == True)
        check('Has tags', 'shreck' in str(cfg.get('profile', {}).get('profile', {}).get('tags', [])))
        check('Markup set', cfg.get('profile', {}).get('markup') == 15)
    except Exception as e:
        check('config.json valid', False, str(e))

    # Verify SOUL.md content
    soul = open(os.path.join(template_dir, 'SOUL.md')).read()
    check('SOUL has role', 'Shreck' in soul)
    check('SOUL has personality', 'Scottish' in soul or 'grump' in soul.lower())
    check('SOUL has catchphrases', 'swamp' in soul.lower())
else:
    check('Template dir created', False, 'not found at ' + template_dir)

# Kill the process (it's probably waiting for setup to proceed)
try:
    p.close(force=True)
except:
    pass

print(f'\n══════════════════════════════════════════════════')
print(f'  Results: {passed} passed, {failed} failed')
print(f'══════════════════════════════════════════════════\n')

sys.exit(1 if failed > 0 else 0)
